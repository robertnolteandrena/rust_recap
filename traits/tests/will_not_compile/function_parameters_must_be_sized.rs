use rand::{thread_rng, Rng};
use spectral::assert_that;
use traits_lib::{Goat, HasName, Sheep};
fn test_function<T: HasName>(t: &T) -> String {
    format!("this is test_function: {}", t.name())
}
fn main() {
    let mut rng = thread_rng();
    let is_goat = rng.gen();
    let name_haver: &dyn HasName = if is_goat { &Goat } else { &Sheep };
    let test_string = test_function(name_haver);
    if is_goat {
        assert_that!(test_string.as_ref()).is_equal_to("this is test_function Goat");
    } else {
        assert_that!(test_string.as_ref()).is_equal_to("this is test_function Sheep");
    }
}
