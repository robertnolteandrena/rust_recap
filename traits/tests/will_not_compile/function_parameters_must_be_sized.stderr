error[E0277]: the size for values of type `dyn HasName` cannot be known at compilation time
  --> tests/will_not_compile/function_parameters_must_be_sized.rs:11:37
   |
11 |     let test_string = test_function(name_haver);
   |                       ------------- ^^^^^^^^^^ doesn't have a size known at compile-time
   |                       |
   |                       required by a bound introduced by this call
   |
   = help: the trait `Sized` is not implemented for `dyn HasName`
note: required by a bound in `test_function`
  --> tests/will_not_compile/function_parameters_must_be_sized.rs:4:18
   |
4  | fn test_function<T: HasName>(t: &T) -> String {
   |                  ^ required by this bound in `test_function`
help: consider relaxing the implicit `Sized` restriction
   |
4  | fn test_function<T: HasName + ?Sized>(t: &T) -> String {
   |                             ++++++++
